### DISCUSSION TITLE
Adding UID to StockItem form?

### DISCUSSION THREAD

**Discussion by dev-bddw** [inventree/InvenTree#2973] [General]
Hello - I'm trying to add UID fields to stock item creation forms in Inventree. I've tried editing CreateStockItemForm class in stock/forms.py and adding uid to the fields list as so, but it doesn't change the form. 

   ` class Meta:
        model = StockItem
        fields = [
            'uid',
            'part',
            'supplier_part',
            'location',
            'quantity',
            'batch',
            'serial_numbers',
            'packaging',
            'purchase_price',
            'expiry_date',
            'link',
            'delete_on_deplete',
            'status',
            'owner',
        ]`

Can anyone help?

**Comment by SchrodingersGat** [comment]
@dev-bddw what version of InvenTree are you using? From 0.6.x onwards, the `CreateStockItemForm` class is actually outdated and doesn't do anything. It has been removed from the master branch.

**Reply by dev-bddw** [reply]
What is the UID field doing on the StockItem? What is it for?

**Reply by SchrodingersGat** [reply]
The UID field is used to store barcode data against the stock item. It should not be edited directly by the user. It is used by barcode plugins to identify the stock item

**Reply by dev-bddw** [reply]
Thanks for your help! How would you handle a case where we wanted to assign a unique ID number to a stock item with quantity greater than 1?

**Reply by dev-bddw** [reply]
We're trying to find an attribute on the StockItem that can only be allowed for a single item. 

For instance, if we had two parts each with a single stock item, we would want this stock item attribute to be unique between the two stock items. Right now they can share a serial number (if the part is different) and it appears that it can also share a UID.

**Reply by SchrodingersGat** [reply]
Each StockItem has a primary key (pk) value which is a numerical value that is guaranteed unique. This PK value is used to uniquely identify each StockItem in the database, and is already available. Can you use that?

![image](https://user-images.githubusercontent.com/10080325/167958765-3e1c487a-64e6-4498-b7eb-9e46f5f9bdc5.png)

**Comment by matmair** [comment]
@dev-bddw we do not recommend changing things in the core code. We release often and merging upstream can quickly become a sizeable job. If you need a unique identifier I would suggest just using the barcode association feature.

**Reply by dev-bddw** [reply]
I will look into this thank you.

**Reply by matmair** [reply]
@dev-bddw please also update to 0.6.x as @SchrodingersGat pointed out. We are only supporting the currently running stable branch (0.6.x)

**Reply by dev-bddw** [reply]
Ok -- I updated, although I'm certain I was using a 0.6.X version of the code.

**Comment by dev-bddw** [comment]
PK value seems very promising, we'll look into that next, thanks! We just
need to be able to manually input that value, we'll dig around a bit.

On Wed, May 11, 2022 at 6:37 PM Oliver ***@***.***> wrote:

> Each StockItem has a primary key (pk) value which is a numerical value
> that is guaranteed unique. This PK value is used to uniquely identify each
> StockItem in the database, and is already available. Can you use that?
>
> [image: image]
> <https://user-images.githubusercontent.com/10080325/167958765-3e1c487a-64e6-4498-b7eb-9e46f5f9bdc5.png>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/inventree/InvenTree/discussions/2973#discussioncomment-2733725>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ARTOWWCHN64BSOFC3TJN63TVJQZDPANCNFSM5VVICRTA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
Jon Benjamin
BDDW
484.868.8116

**Comment by dev-bddw** [comment]
We’ll update to 0.6.x before we get too far into this, but just at first
glance, getting a sense of how accessible everything is - the default stock
CRUD has no way of overriding or even noting the pk number (aside from the
stock item’s url), and even the inventree admin backend view for stock
items has no way of dealing with pk values, so I’m guessing this may be a
bit of a locked box for the system? Aka, no way to get in and edit those
values?


On Thu, May 12, 2022 at 9:13 AM Jon Benjamin ***@***.***> wrote:

> PK value seems very promising, we'll look into that next, thanks! We just
> need to be able to manually input that value, we'll dig around a bit.
>
> On Wed, May 11, 2022 at 6:37 PM Oliver ***@***.***> wrote:
>
>> Each StockItem has a primary key (pk) value which is a numerical value
>> that is guaranteed unique. This PK value is used to uniquely identify each
>> StockItem in the database, and is already available. Can you use that?
>>
>> [image: image]
>> <https://user-images.githubusercontent.com/10080325/167958765-3e1c487a-64e6-4498-b7eb-9e46f5f9bdc5.png>
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/inventree/InvenTree/discussions/2973#discussioncomment-2733725>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/ARTOWWCHN64BSOFC3TJN63TVJQZDPANCNFSM5VVICRTA>
>> .
>> You are receiving this because you were mentioned.Message ID:
>> ***@***.***>
>>
>
>
> --
> Jon Benjamin
> BDDW
> 484.868.8116
>
-- 
Jon Benjamin
BDDW
484.868.8116

**Reply by SchrodingersGat** [reply]
Correct, the PK value is not editable.
